lib/rpush/daemon/apns2/delivery.rb

```diff
@@ -8,124 +8,132 @@ module Apns2
       class Delivery < Rpush::Daemon::Delivery
         RETRYABLE_CODES = [ 429, 500, 503 ]
 
-        def initialize(app, http2_client, batch, notification)
+        def initialize(app, http2_client, batch)
           @app = app
           @client = http2_client
           @batch = batch
-          @notification = notification
-          @response = OpenStruct.new
         end
 
         def perform
-          do_async_post
+          @batch.each_notification do |notification|
+            prepare_async_post(notification)
+          end
+
+          # Send all preprocessed requests at once
+          @client.join
+
+          @batch.mark_all_delivered
         rescue SocketError => error
-          mark_retryable(@notification, Time.now + 10.seconds, error)
+          mark_batch_retryable(Time.now + 10.seconds, error)
           raise
         rescue StandardError => error
-          mark_failed(error)
+          mark_batch_failed(error)
           raise
         ensure
-          @batch.notification_processed
+          @batch.all_processed
         end
 
         protected
         ######################################################################
 
-        def handle_response
-          code = @response.code
-          case code
-          when 200
-            ok
-          when *RETRYABLE_CODES
-            service_unavailable
-          else
-            reflect(:notification_id_failed,
-              @app,
-              @notification.id, code,
-              @response.failure_reason)
-            fail Rpush::DeliveryError.new(@response.code,
-              @notification.id, @response.failure_reason)
-          end
-        end
-
-        def ok
-          log_info("#{@notification.id} sent to #{@notification.device_token}")
-          mark_delivered
-        end
-
-        def service_unavailable
-          mark_retryable(@notification, Time.now + 10.seconds)
-          # Logs should go last as soon as we need to initialize
-          # retry time to display it in log
-          failed_message_to_log
-          retry_message_to_log
-        end
+        def prepare_async_post(notification)
+          response = OpenStruct.new
```

Comment:
Performance wise I think it would be better use a Struct or Hash here as OpenStruct uses method_missing internally, see http://ruby-doc.org/stdlib-1.9.3/libdoc/ostruct/rdoc/OpenStruct.html#class-OpenStruct-label-Implementation-3A